src/braille.jl

```diff
@@ -8,37 +11,42 @@ Binary dither with algorithm `alg`, then print image in braille.
 
 ## Keyword arguments
   - `invert`: invert Unicode output, defaults to `false`
+  - `to_string`: return a string instead of printing to terminal. Defaults to `false`.
 
 All keyword arguments for binary dithering methods can be used.
 """
-function braille(img::GenericImage, alg::AbstractDither; invert=false, kwargs...)
+function braille(
+    img::GenericImage,
+    alg::AbstractDither;
+    invert::Bool=false,
+    to_string::Bool=false,
+    kwargs...,
+)
     d = dither(Bool, Gray.(img), alg; kwargs...)
-    invert && return brailleprint(.!d)
-    return brailleprint(d)
+    return braille(d; invert=invert, to_string=to_string)
 end
 
-"""
-    brailleprint(img)
+function braille(A::AbstractMatrix{Bool}; invert::Bool=false, to_string::Bool=false)
+    invert && (A .= .!A)
+    B = _braille_matrix(A)
+    str = join([join(r) for r in eachrow(B)], "\n")
+    to_string && return str
+    print(str)
+    return nothing
+end
 
-Interpret binary image `img` as sparse array and print it in braille through UnicodePlots.
-Refer to the UnicodePlots.jl documentation of `spy` for plot arguments.
-"""
-function brailleprint(
-    img::GenericGrayImage; title="", margin=0, padding=0, border=:none, kwargs...
-)
-    h, w = size(img)
-    img = Bool.(img)
-
-    return UnicodePlots.spy(
-        img;
-        # One braille character corresponds to a 4x2 grid
-        maxheight=ceil(Int, h / 4),
-        maxwidth=ceil(Int, w / 2),
-        title=title,
-        margin=margin,
-        padding=padding,
-        border=border,
-        labels=false,
-        kwargs...,
-    )
+_braille_index(A) = sum(view(A, :) .* BRAILLE_CODE) + 1
```

Comment:
`view(A, :) .* BRAILLE_CODE` repeatly allocate memories...

```suggestion
Base.@propagate_inbounds function _braille_index(A)
    idx = 1
    for i in 1:length(A)
        idx += A[i] * BRAILLE_CODE[i]
    end
    return idx
end
```